crates/rolldown_plugin_vite_resolve/src/vite_resolve_plugin.rs

```diff
@@ -0,0 +1,402 @@
+  ) -> HookResolveIdReturn {
+    if args.specifier.starts_with('\0')
+      || args.specifier.starts_with("virtual:")
+      || args.specifier.starts_with("/virtual:")
+    {
+      return Ok(None);
+    }
+
+    if args.specifier.starts_with(BROWSER_EXTERNAL_ID) {
+      // TODO: implement for dev
+      return Ok(Some(HookResolveIdOutput {
+        id: args.specifier.to_string(),
+        ..Default::default()
+      }));
+    }
+
+    if self.resolve_options.as_src && args.specifier.starts_with(FS_PREFIX) {
+      // TODO: implement for dev
+      let res = fs_path_from_id(args.specifier);
+      return Ok(Some(HookResolveIdOutput { id: res.to_string(), ..Default::default() }));
+    }
+
+    if args.specifier.starts_with("file://") {
+      // TODO: implement fileURLToPath properly
+      let mut res = args.specifier.replace("file://", "");
+      if res.starts_with('/') && is_windows_drive_path(&res[1..]) {
+        res.remove(0);
+      }
+      return Ok(Some(HookResolveIdOutput { id: res, ..Default::default() }));
+    }
+
+    if args.specifier.trim_start().starts_with("data:") {
+      return Ok(None);
+    }
+
+    if is_external_url(args.specifier) {
+      return Ok(Some(HookResolveIdOutput {
+        id: args.specifier.to_string(),
+        external: Some(true),
+        ..Default::default()
+      }));
+    }
+
+    let additional_options = AdditionalOptions::new(
+      args.kind == ImportKind::Require,
+      self.resolve_options.prefer_relative || args.importer.map_or(false, |i| i.ends_with(".html")),
+      is_from_ts_importer(args.importer),
+    );
+    let resolver = self.resolver.get(additional_options);
+
+    if is_bare_import(args.specifier) {
+      let base_dir = if let Some(importer) = args.importer {
+        let imp = Path::new(importer);
+        if imp.is_absolute()
+          && (importer.ends_with('*') || fs::exists(clean_url(importer)).unwrap_or(false))
+        {
+          imp.parent().map(|i| i.to_str().unwrap()).unwrap_or(importer)
+        } else {
+          &self.resolve_options.root
+        }
+      } else {
+        &self.resolve_options.root
+      };
+
+      let oxc_resolved_result = resolver.resolve(base_dir, args.specifier);
+      let resolved = normalize_oxc_resolver_result(&self.resolver, &oxc_resolved_result)?;
+      if let Some(mut resolved) = resolved {
+        // TODO: handle shouldExternal
+        let external = self.environment_consumer == "server";
+
+        if external {
+          let resolved_ext = get_extension(&resolved.id);
+          if resolved_ext == "js" || resolved_ext == "mjs" || resolved_ext == "cjs" {
+            let id = args.specifier;
+            let mut resolved_id = id;
+            if is_deep_import(id) && resolved_ext != get_extension(id) {
+              if let Some(pkg_json) = oxc_resolved_result.unwrap().package_json() {
+                let has_exports_field =
+                  pkg_json.raw_json().as_object().unwrap().get("exports").is_some();
```

Comment:
@Boshen Does it make sense to have a feature to check the existence of `exports` field in `oxc_resolver`? This is the only reason I'm enabling `package_json_raw_json_api` feature. or is it fine to rely on `package_json_raw_json_api` feature?

This code is adding needed extensions for imports pointing external packages.
For example, you have `import foo from 'pkg/a'` in your code and `pkg` has `a.js` and doesn't have `exports` field. In the application code base, you can write `pkg/a` because Vite allows omitting extensions. But after the bundle that import needs to be `pkg/a.js` because Node.js does not allow omitting extensions in ESM. This code transforms `pkg/a` to `pkg/a.js` so that users can write `pkg/a` for imports pointing external packages.
Maybe it make sense to have this whole feature on the oxc-resolver side?